src/ui/public/directives/match_pairs.js

```diff
@@ -0,0 +1,101 @@
+import { uiModules } from 'ui/modules';
+const module = uiModules.get('kibana');
+
+/**
+ * This directively automatically handles matching pairs.
+ * Specifically, it does the following:
+ *
+ * 1. If the key is a closer, and the character in front of the cursor is the
+ *    same, simply move the cursor forward.
+ * 2. If the key is an opener, insert the opener at the beginning of the
+ *    selection, and the closer at the end of the selection, and move the
+ *    selection forward.
+ * 3. If the backspace is hit, and the characters before and after the cursor
+ *    are a pair, remove both characters and move the cursor backward.
+ */
+
+const pairs = ['()', '[]', '{}', `''`, '""'];
+const openers = pairs.map(pair => pair[0]);
+const closers = pairs.map(pair => pair[1]);
+
+module.directive('matchPairs', () => ({
+  restrict: 'A',
+  require: 'ngModel',
+  link: function (scope, elem, attrs, ngModel) {
+    elem.on('keydown', (e) => {
+      const { target, key, metaKey } = e;
+      const { value, selectionStart, selectionEnd } = target;
+
+      if (shouldMoveCursorForward()) {
```

Comment:
I hate to be that guy, but you should probably [return early](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#returnthrow-early-from-functions).

```js
if (shouldMoveCursorForward()) return moveCursorForward();
if (shouldInsertMatchingCloser()) return insertMatchingCloser();
if (shouldRemovePair()) return removePair();
```